adlik_serving/runtime/tensorrt/model/trt_instance.cc

```diff
@@ -79,67 +101,32 @@ struct Instance : BatchProcessor {
   const int max_batch_size;
 
   // TensorRT components for the model
-  nvinfer1::IRuntime* runtime;
-  nvinfer1::ICudaEngine* engine;
-  nvinfer1::IExecutionContext* context;
+  TrtPtr<nvinfer1::IRuntime> runtime;
+  TrtPtr<nvinfer1::ICudaEngine> engine;
+  TrtPtr<nvinfer1::IExecutionContext> context;
 
   // The number of inputs required for this model.
   size_t num_inputs;
 
   // For each binding index of the TensorRT engine, the size of the
   // corresponding tensor and pointer to the CUDA buffer for the
   // tensor. These are arrays with size equal to number of bindings.
-  uint64_t* byte_sizes;
-  void** buffers;
+  std::unique_ptr<uint64_t[]> byte_sizes;
+  std::unique_ptr<void* []> buffers;
 
   // The stream where operations are executed.
-  cudaStream_t stream;
+  std::unique_ptr<CUstream_st, StreamDeleter> stream;
 };
 ////////////////////////////////////////////////////////////////////////////////////
 
 Instance::Instance(const ModelConfigProto& config, const std::string& name, const int gpu_device)
-    : config(config),
-      name(name),
-      gpu_device(gpu_device),
-      max_batch_size(config.max_batch_size()),
-      runtime(nullptr),
-      engine(nullptr),
-      context(nullptr),
-      num_inputs(0),
-      byte_sizes(nullptr),
-      buffers(nullptr),
-      stream(nullptr) {
-}
-
-Instance::Instance(Instance&& obj)
-    : config(obj.config),
-      name(std::move(obj.name)),
-      gpu_device(obj.gpu_device),
-      max_batch_size(obj.max_batch_size),
-      runtime(obj.runtime),
-      engine(obj.engine),
-      context(obj.context),
-      num_inputs(obj.num_inputs),
-      byte_sizes(obj.byte_sizes),
-      buffers(obj.buffers),
-      stream(obj.stream) {
-  obj.runtime = nullptr;
-  obj.engine = nullptr;
-  obj.context = nullptr;
-  obj.num_inputs = 0;
-  obj.byte_sizes = nullptr;
-  obj.buffers = nullptr;
-  obj.stream = nullptr;
-}
+    : config(config), name(name), gpu_device(gpu_device), max_batch_size(config.max_batch_size()), num_inputs(0) {
+}  // namespace
 
 Instance::~Instance() {
   INFO_LOG << "~Instance::Instance ";
 
-  if (byte_sizes != nullptr) {
-    delete[] byte_sizes;
-    byte_sizes = nullptr;
-  }
-  if (buffers != nullptr) {
+  if (buffers) {
```

Comment:
Could `buffers`  be represented as as a instance of `std::unique_ptr<std::unique_ptr<void, CudaDeleter>[]>`, where:

```c++
struct CudaDeleter {
    void operator()(void* p) const {
        std::cout << "I am going to destroy you, " << p << '.' << std::endl;
    }
};
```